### PR TITLE
Jw flipped lidar

### PR DESCRIPTION
**Pull request by Vuwij**
Adds an option to flip the lidar

### CODE REVIEW

File: flatland_plugins/src/laser.cpp [hunk 1]
```diff
@@ -191,8 +191,16 @@ void Laser::ComputeLaserRanges() {
   // Unqueue all of the future'd results
   for (unsigned int i = 0; i < laser_scan_.ranges.size(); ++i) {
     auto result = results[i].get();  // Pull the result from the future
-    laser_scan_.ranges[i] = result.first + this->noise_gen_(this->rng_);
-    if (reflectance_layers_bits_) laser_scan_.intensities[i] = result.second;
+
+    if (!flipped_) {
```
Comment (hendrik-skubch) [inline flatland_plugins/src/laser.cpp]: Positive case first.

File: flatland_plugins/src/laser.cpp [hunk 2]
```diff
@@ -191,8 +191,16 @@ void Laser::ComputeLaserRanges() {
...
+      laser_scan_.ranges[i] = result.first + this->noise_gen_(this->rng_);
+      if (reflectance_layers_bits_) laser_scan_.intensities[i] = result.second;
```
Comment (hendrik-skubch) [inline flatland_plugins/src/laser.cpp]: braces

**Review by hendrik-skubch — COMMENTED** [review]
Err, why can this not be handled through the urdf?

File: flatland_server/src/debug_visualization.cpp [hunk 3]
```diff
@@ -156,18 +156,26 @@ void DebugVisualization::BodyToMarkers(visualization_msgs::MarkerArray& markers,
 
       } break;
 
-      case b2Shape::e_polygon: {  // Convert b2Polygon -> LINE_STRIP
+      case b2Shape::e_polygon: {  // Currently only
```
Comment (gauthamm) [inline flatland_server/src/debug_visualization.cpp]: currently only?? what does that mean?

File: flatland_plugins/src/laser.cpp (hunk 1 above)
Comment (gauthamm) [inline flatland_plugins/src/laser.cpp]:
```suggestion
    unsigned int index = (flipped_)? (laser_scan_.ranges.size() - i) : i;
    laser_scan_.ranges[index] = result.first + this->noise_gen_(this->rng_);	
    if (reflectance_layers_bits_) {
         laser_scan_.intensities[index] = result.second;
    }
```

File: flatland_plugins/src/laser.cpp [hunk 4]
```diff
@@ -191,8 +191,17 @@ void Laser::ComputeLaserRanges() {
...
+    int index;
+    if (flipped_) {
+      index = laser_scan_.ranges.size() - i;
+    } else {
+      index = i;
+    }
+
+    laser_scan_.ranges[index] = result.first + this->noise_gen_(this->rng_);
+    if (reflectance_layers_bits_)
+      laser_scan_.intensities[index] = result.second;
```
Comment (gauthamm) [inline flatland_plugins/src/laser.cpp]: still missing braces

**Review by gauthamm — CHANGES_REQUESTED** [review]

File: flatland_plugins/src/laser.cpp [hunk 5]
```diff
@@ -191,8 +191,17 @@ void Laser::ComputeLaserRanges() {
   // Unqueue all of the future'd results
   for (unsigned int i = 0; i < laser_scan_.ranges.size(); ++i) {
     auto result = results[i].get();  // Pull the result from the future
-    laser_scan_.ranges[i] = result.first + this->noise_gen_(this->rng_);
-    if (reflectance_layers_bits_) laser_scan_.intensities[i] = result.second;
+
+    int index;
+    if (flipped_) {
```
Comment (hendrik-skubch) [inline flatland_plugins/src/laser.cpp:203]: having this check outside the loop is better imo. Now it is very hard for the compiler to understand that this is linear loop. You should be able to cast this as a use-case of std::transform.
Something along the lines of
`std::transform(results.begin(),results.end(),(_flipped? laser_scan_.ranges.rbegin() :  laser_scan_.ranges.begin(), [](const auto& res) { ... } );`

**Review by gauthamm — APPROVED** [review]

**Pull request merged**